### PR TITLE
cache sigar probes to prevent incorrect calculation of percentual values

### PR DESCRIPTION
**Pull request by chaudum**
for example, statements such as:

```sql
SELECT id, load, os['cpu'] FROM sys.nodes ORDER BY id;
```

lead to incorrect values of ``os['cpu']`` (returning only ``0`` and ``100``)
this was due to calling sigar in too quick succession and therefore not
being able to calculate percentual values based on the short time delta
between the calls.

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -5,6 +5,10 @@ Changes for Crate
 Unreleased
 ==========
 
+ - Fix: selecting ``os``, ``network``, ``fs`` or ``process`` from ``sys.nodes``
+   in very quick succession or multiple times in same select statement
+   could lead to wrong results for percentual values from sigar
```
Comment (seut) [inline CHANGES.txt]: suggest: "from the underlying sigar library." (also missing . at the end of the sentence)

File: sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java [hunk 2]
```diff
@@ -33,15 +35,31 @@
...
+        this.processCpuStatsCache = new ExtendedProcessCpuStatsCache(PROBE_CACHE_TIME, processCpuStatsProbe());
     }
 
     @Override
-    public ExtendedNetworkStats networkStats() {
+    public synchronized ExtendedNetworkStats networkStats() {
+        return networkStatsCache.getOrRefresh();
+    }
+
+    public ExtendedNetworkStats networkStatsProbe() {
```
Comment (seut) [inline sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java]: make private

File: sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java [hunk 3]
```diff
@@ -33,15 +35,31 @@
...
+        this.nodeEnvironment = nodeEnvironment;
+        this.osStatsCache = new ExtendedOsStatsCache(PROBE_CACHE_TIME, osStatsProbe());
+        this.fsStatsCache = new ExtendedFsStatsCache(PROBE_CACHE_TIME, fsStatsProbe());
+        this.networkStatsCache = new ExtendedNetworkStatsCache(PROBE_CACHE_TIME, networkStatsProbe());
+        this.processCpuStatsCache = new ExtendedProcessCpuStatsCache(PROBE_CACHE_TIME, processCpuStatsProbe());
     }
 
     @Override
-    public ExtendedNetworkStats networkStats() {
+    public synchronized ExtendedNetworkStats networkStats() {
```
Comment (seut) [inline sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java]: as the SingleObjectCache uses a lock to protect against concurrent calls, synchronized is not needed here afaik (same for all other methods)

**Comment by seut** [comment]
if possible, a test for this issue would be great

File: sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java (hunk 3 above)
Comment (chaudum) [inline sigar/src/main/java/io/crate/monitor/SigarExtendedNodeInfo.java]: I took it from there: https://github.com/elastic/elasticsearch/blob/1.7/src/main/java/org/elasticsearch/monitor/os/OsService.java#L57

**Comment by chaudum** [comment]
superseded by https://github.com/crate/crate/pull/3755

**Pull request closed without merging**